### PR TITLE
Fix sass types for TS 4.7+ Node16/NodeNext module resolution

### LINKED ISSUE #1714: TypeScript types not accessible with new TS 4.7 Node16/next resolution
As of TypeScript 4.7, the `exports` field can be respected when the new resolution (`node16` or `nodenext`) is used. The Sass npm package uses `exports` but does not set the `types` condition. This results in `d.ts` resolution errors like the following:

```
Could not find a declaration file for module 'sass'. '/X/node_modules/sass/sass.default.dart.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/sass` if it exists or add a new declaration (.d.ts) file containing `declare module 'sass';`

3 import sass from 'sass';
```

One might wonder why this happens when the `sass/package.json` sets the `types` field. This field is ignored when the `exports` are configured and the `default` condition is pointing to the `.js` file, preventing `.d.ts` from being discovered.

The fix is to add another `types` condition that points to the index `.d.ts` file. This seems to require changing `dart_cli_pkg` to either add this per config-setting, or to merge the `exports` from the user-provided "base package.json` file.

### PR DESCRIPTION
**Pull request by Goodwine**
Fixes #1714

### CODE REVIEW

File: package/package.json [hunk 1]
```diff
@@ -21,6 +21,18 @@
     "immutable": "^4.0.0",
     "source-map-js": ">=0.6.2 <2.0.0"
   },
-  "keywords": ["style", "scss", "sass", "preprocessor", "css"],
-  "types": "types/index.d.ts"
+  "devDependencies": {
+    "@types/node": "^18.0.0"
+  },
```
Comment (nex3) [inline package/package.json:26]: `devDependencies` doesn't make sense—those aren't used by packages that depend on this package, and this `package.json` isn't ever used for local development. I _think_ this should just be normal `dependencies`.

**Review by nex3 — CHANGES_REQUESTED** [review]

File: package/package.json (hunk 1 above)
Comment (ntkme) [reply package/package.json]: @nex3 @Goodwine I don't think this should be added at as `dependencies`, because it is the user's choice to use TS or JS. When using pure JS, we shouldn't need any types to be installed as a dependency.If this is affecting the TS compilation of this project, it should indeed be a `devDependencies`.

Comment (Goodwine) [reply package/package.json]: Wouldn't the `@types/node` package be downloaded regardless of whether it being declared as `dependencies` vs `devDependencies`? or do devDependencies get skipped if `sass` gets declared as non-dev dependency?

Comment (nex3) [reply package/package.json]: > do devDependencies get skipped if sass gets declared as non-dev dependency?

This is correct, unless I'm severely misunderstanding. Sometimes it's necessary to include a dependency that's not used by some users because it's necessary for others.

**Review by nex3 — APPROVED** [review]

File: package/package.json (hunk 1 above)
Comment (devversion) [reply package/package.json]: FWIW: Node packages (even when shipping `.d.ts`) usually never add a dependency to satisfy TypeScript type-checking for users. Instead:

* Users either have to set up `@types/node` themselves, or the library authors place a directive implying the necessary type like: `/// <reference types="node" />` into their `.d.ts`
* Or, users most of the time just disable type-checking of external libraries. e..g by setting `skipLibCheck: true`.

Also just to make this clear: The `devDependencies` will never be installed for users, only for the authors of the package when e.g. `yarn`/`npm` is executed inside the folder containing the `package.json`.

TL;DR: Adding `@types/node` as a dependency is pretty uncommon. It should be removed. It's also not a big deal though and doesn't impact anyone negatively.

**Comment by devversion** [comment]
Thanks for fixing this! I just added a little comment above (which you may want to consider, or not)

File: package/package.json (hunk 1 above)
Comment (nex3) [reply package/package.json]: If including `@types/node` doesn't impact anyone negatively, and means that users can freely use the TS types without needing to manually install additional dependencies, why _not_ include it?

Comment (ntkme) [reply package/package.json]: Well, this may become an addition to the reasons we get bloated node_modules.

Also, having “@types/node” lock to version 18 is not really a good idea given that end users might want to install other versions (e.g. 16) if that is the version of node they are using.

Comment (devversion) [reply package/package.json]: Mainly for good practice of not increasing the dependency tree for people not using the TS types, which is very likely the majority here. This is peanuts in size, but a general question of best practice.

You also don't have a peer dependency/ or dependency on TypeScript itself, making guarantees with what version your types work 🧐)

I don't care personally, but this is uncommon. Angular for example doesn't do it.

Comment (nex3) [reply package/package.json]: > Also, having `@types/node` lock to version 18 as part of `dependencies` is not really a good idea given that end users might want to install other versions (e.g. 16) if that is the version of node they are using.

This is a good point—we should look into making the version range here as wide as possible while still including the APIs we use (I think only `Buffer`?).

> You also don't have a peer dependency/ or dependency on TypeScript itself, making guarantees with what version your types work

Ideally, we'd be able to specify a dependency of the form "*if* the downstream user has TypeScript installed, *then* it must have this version", but unfortunately peer dependencies will produce warnings even if the downstream user doesn't have TypeScript.
